### DISCUSSION TITLE
Supporting custom types

### DISCUSSION THREAD

**Discussion by WoobaWooba** [OAI/OpenAPI-Specification#2944] [Q&A]
Hi,

I'm new to OpenAPI and we're using for a project where we have our custom data types. 

I would like to know if it would be possible to use an UI implementation of the OpenAPI and
have it render the custom data type. For example, I could use this json,

`{
    "varName":{
       "type": "customDataType"
    }
}`

and have it rendered as it it were a string or number or any of the other supported types by OA.

P.S.: We could use objects but we would still have to fall back to the OpenAPI data types, which is something we want to avoid.

Thanks

**Comment by flaksp** [comment]
OpenAPI 3.1 or newer uses [JSON Schema (draft 2020-12)](https://json-schema.org/specification-links.html#2020-12) spec to define schemas.

The `type` keyword in JSON Schema defines name of the type that may appear in JSON, this is [what the spec says](https://json-schema.org/draft/2020-12/json-schema-validation.html#section-6.1.1):

> The value of this keyword MUST be either a string or an array. If it is an array, elements of the array MUST be strings and MUST be unique.
>
> String values MUST be one of the six primitive types ("null", "boolean", "object", "array", "number", or "string"), or "integer" which matches any number with a zero fractional part.

If `type` keyword will contain something unexpected, it will lead to invalid JSON Schema, so existing tools won't understand it. This is because in JSON we don't have types named `customDataType` from your example, we have only nulls, booleans, objects, arrays, numbers and strings.

If you want to specify some data type that has some semantics behind it, you may want to use [`format`](https://json-schema.org/draft/2020-12/json-schema-validation.html#section-7) keyword. It may be used to describe dates, emails, IP addresses, URIs, etc. Full list of pre-defined data types available in [7.3 "Defined Formats" section of the JSON Schema Validation spec](https://json-schema.org/draft/2020-12/json-schema-validation.html#section-7.3). OpenAPI also defines its own set of pre-defined data types in [4.4 "Data Types" section of the spec](https://spec.openapis.org/oas/latest.html#data-types).

If pre-defined data types are not enough for you, [you are allowed to use your own data types](https://json-schema.org/draft/2020-12/json-schema-validation.html#section-7.2.3), but there are no guarantees implementations you use will support them as you expect.

Valid example of a schema with `format` keyword:

```json
{
  "type": "string",
  "format": "email"
}
```

It's also possible to reuse schemas with [`$ref`](https://json-schema.org/draft/2020-12/json-schema-core.html#section-8.2.3) keyword.

You may define your schema once in [Components Object](https://spec.openapis.org/oas/latest.html#components-object):

```json
{
  "openapi": "3.1.0",
  "components": {
    "schemas": {
      "Email": {
        "type": "string",
        "format": "email"
      }
    }
  }
}
```

And then reuse it inside your OpenAPI document with `$ref`:

```json
{
  "primaryEmail": { "$ref": "#/components/schemas/Email" }
}
```

**Comment by handrews** [comment]
This was answered, and `type` is governed by JSON Schema rather than OpenAPI.  We definitely will not be introducing any incompatibilities involving such a core keyword, especially after going to great lengths to resolve all JSON Schema incompatibilities in OAS 3.1.  Closing.
